match.go

```diff
@@ -0,0 +1,493 @@
+//
+// Derived from github.com/maxatome/go-testdeep/internal/trace.SplitPackageFunc.
+func extractPackage(fn string) string {
+	sp := strings.LastIndexByte(fn, '/')
+	if sp < 0 {
+		sp = 0 // std package
+	}
+
+	dp := strings.IndexByte(fn[sp:], '.')
+	if dp < 0 {
+		return ""
+	}
+
+	return fn[:sp+dp]
+}
+
+// calledFrom returns a string like "@PKG.FUNC() FILE:LINE".
+func calledFrom(skip int) string {
+	pc := make([]uintptr, 128)
+	npc := runtime.Callers(skip+1, pc)
+	pc = pc[:npc]
+
+	frames := runtime.CallersFrames(pc)
+
+	var lastFrame runtime.Frame
+
+	for {
+		frame, more := frames.Next()
+
+		// If testing package is encountered, it is too late
+		if strings.HasPrefix(frame.Function, "testing.") {
+			break
+		}
+		lastFrame = frame
+		// Stop if httpmock is not the caller
+		if !ignorePackages[extractPackage(frame.Function)] || !more {
+			break
+		}
+	}
+
+	if lastFrame.Line == 0 {
+		return ""
+	}
+	return fmt.Sprintf(" @%s() %s:%d",
+		lastFrame.Function, lastFrame.File, lastFrame.Line)
+}
+
+// MatcherFunc type is the function to use to check a Matcher matches
+// an incoming request.
+type MatcherFunc func(req *http.Request) bool
+
+func matcherFuncOr(mfs []MatcherFunc) MatcherFunc {
+	return func(req *http.Request) bool {
+		for _, mf := range mfs {
+			if mf(req) {
+				return true
+			}
+		}
+		return false
+	}
+}
+
+func matcherFuncAnd(mfs []MatcherFunc) MatcherFunc {
+	if len(mfs) == 0 {
+		return nil
+	}
+	return func(req *http.Request) bool {
+		for _, mf := range mfs {
+			if !mf(req) {
+				return false
+			}
+		}
+		return true
+	}
+}
+
+// Check returns true if mf is nil, otherwise it returns mf(req).
+func (mf MatcherFunc) Check(req *http.Request) bool {
+	return mf == nil || mf(req)
+}
+
+// Or combines mf and all mfs in a new [MatcherFunc]. This new
+// [MatcherFunc] succeeds if one of mf or mfs succeeds. Note that as a
+// a nil [MatcherFunc] is considered succeeding, if mf or one of mfs
+// items is nil, nil is returned.
+func (mf MatcherFunc) Or(mfs ...MatcherFunc) MatcherFunc {
+	if len(mfs) == 0 || mf == nil {
+		return mf
+	}
+	cmfs := make([]MatcherFunc, len(mfs)+1)
+	cmfs[0] = mf
+	for i, cur := range mfs {
+		if cur == nil {
+			return nil
+		}
+		cmfs[i+1] = cur
+	}
+	return matcherFuncOr(cmfs)
+}
+
+// And combines mf and all mfs in a new [MatcherFunc]. This new
+// [MatcherFunc] succeeds if all of mf and mfs succeed. Note that a
+// [MatcherFunc] also succeeds if it is nil, so if mf and all mfs
+// items are nil, nil is returned.
+func (mf MatcherFunc) And(mfs ...MatcherFunc) MatcherFunc {
+	if len(mfs) == 0 {
+		return mf
+	}
+	cmfs := make([]MatcherFunc, 0, len(mfs)+1)
+	if mf != nil {
+		cmfs = append(cmfs, mf)
+	}
+	for _, cur := range mfs {
+		if cur != nil {
+			cmfs = append(cmfs, cur)
+		}
+	}
+	return matcherFuncAnd(cmfs)
+}
+
+// Matcher type defines a match case. The zero Matcher{} corresponds
+// to the default case. Otherwise, Name and Matcher can never have the
```

Comment:
```suggestion
// to the default case. Otherwise, use [NewMatcher] or any helper
// building a [Matcher] like [BodyContainsBytes], [BodyContainsBytes],
// [HeaderExists], [HeaderIs], [HeaderContains] or any of
// [github.com/maxatome/tdhttpmock] functions.
```